fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneRecordCursor.java

```diff
@@ -0,0 +1,237 @@
+                }
+            }, executor);
+        } else { // a limit was exceeded
+            if (limitRemaining <= 0) {
+                nextResult = RecordCursorResult.withoutNextValue(continuationHelper(), NoNextReason.RETURN_LIMIT_REACHED);
+            } else if (currentPosition >= topDocs.scoreDocs.length) {
+                nextResult = RecordCursorResult.withoutNextValue(continuationHelper(), NoNextReason.SOURCE_EXHAUSTED);
+            } else {
+                final Optional<NoNextReason> stoppedReason = limitManager.getStoppedReason();
+                if (!stoppedReason.isPresent()) {
+                    throw new RecordCoreException("limit manager stopped LuceneRecordCursor but did not report a reason");
+                } else {
+                    nextResult = RecordCursorResult.withoutNextValue(continuationHelper(), stoppedReason.get());
+                }
+            }
+            return CompletableFuture.completedFuture(nextResult);
+        }
+    }
+
+    @Nonnull
+    private RecordCursorContinuation continuationHelper() {
+        if (currentPosition >= topDocs.scoreDocs.length && limitRemaining > 0) {
+            return ByteArrayContinuation.fromNullable(null);
+        } else {
+            return ByteArrayContinuation.fromNullable(Ints.toByteArray(currentPosition));
+        }
+    }
+
+    @Nonnull
+    @Override
+    @Deprecated
+    public CompletableFuture<Boolean> onHasNext() {
+        if (hasNextFuture == null) {
+            hasNextFuture = onNext().thenApply(RecordCursorResult::hasNext);
+        }
+        return hasNextFuture;
+    }
+
+    @Nullable
+    @Override
+    @Deprecated
+    public IndexEntry next() {
+        if (!hasNext()) {
+            throw new NoSuchElementException();
+        }
+        hasNextFuture = null;
+        return nextResult.get();
+    }
+
+    @Nullable
+    @Override
+    @Deprecated
+    public byte[] getContinuation() {
+        return Ints.toByteArray(currentPosition);
+    }
+
+    @Nonnull
+    @Override
+    @Deprecated
+    public NoNextReason getNoNextReason() {
+        return nextResult.getNoNextReason();
+    }
+
+    @Override
+    public void close() {
+        if (indexReader != null) {
+            IOUtils.closeWhileHandlingException(indexReader);
+        }
+        if (hasNextFuture != null) {
+            hasNextFuture.cancel(false);
+        }
+    }
+
+    @Nonnull
+    @Override
+    public Executor getExecutor() {
+        return executor;
+    }
+
+    @Override
+    public boolean accept(@Nonnull RecordCursorVisitor visitor) {
+        visitor.visitEnter(this);
+        return visitor.visitLeave(this);
+    }
+
+    private synchronized IndexReader getIndexReader() throws IOException {
+        IndexWriterCommitCheckAsync writerCheck = getIndexWriterCommitCheckAsync(state);
+        return writerCheck == null ? DirectoryReader.open(getOrCreateDirectoryCommitCheckAsync(state).getDirectory()) : DirectoryReader.open(writerCheck.indexWriter);
```

Comment:
I'm assuming this is to deal with this note from the [lucene javadoc](https://lucene.apache.org/core/7_7_2/core/org/apache/lucene/store/Directory.html)
> A file open for writing may not be available for read access until the corresponding IndexOutput is closed.

That being said, will this work correctly if the writer is created after starting the cursor. Eg you wanted to do something like an update, where you pipe the result into code that updates records (e.g. I want to update a typo in all the docs)